torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
                     else:
                         p_assert(
-                            not p._is_sharded, "All sharded parameters should "
+                            not p._is_sharded or not getattr(p, '_post_backward_called', False), "All sharded parameters that received gradient should "
                             "use `_saved_grad_shard`"
                         )
                     if hasattr(p, "_saved_grad_shard"):
                         delattr(p, "_saved_grad_shard")
+                    if hasattr(p, '_post_backward_called'):
```

Comment:
This seems like a no-op because `_post_backward_called` will have already been deleted before `_finalize_params()` is called.